network/p2p/inspector/validation/control_message_validation.go

```diff
@@ -307,12 +328,94 @@ func (c *ControlMsgValidationInspector) validateTopics(ctrlMsgType p2p.ControlMe
 	return nil
 }
 
-// validateTopic the topic is a valid flow topic/channel.
-// All errors returned from this function can be considered benign.
-func (c *ControlMsgValidationInspector) validateTopic(topic channels.Topic) error {
+// validateTopic ensures the topic is a valid flow topic/channel.
+// Expected error returns during normal operations:
+//   - channels.ErrInvalidTopic: if topic is invalid.
+//   - ErrActiveClusterIDsNotSet: if the cluster ID provider is not set.
+//   - ErrActiveClusterIDS: if an error is encountered while getting the active cluster IDs list. This error indicates an unexpected bug or state corruption.
+//   - channels.ErrUnknownClusterID: if the topic contains a cluster ID prefix that is not in the active cluster IDs list.
+//
+// This func returns an exception in case of unexpected bug or state corruption if cluster prefixed topic validation
+// fails due to unexpected error returned when getting the active cluster IDS.
+func (c *ControlMsgValidationInspector) validateTopic(from peer.ID, topic channels.Topic) error {
+	channel, ok := channels.ChannelFromTopic(topic)
+	if !ok {
+		return channels.NewInvalidTopicErr(topic, fmt.Errorf("failed to get channel from topic"))
+	}
+
+	// handle cluster prefixed topics
+	if channels.IsClusterChannel(channel) {
+		return c.validateClusterPrefixedTopic(from, topic)
+	}
+
+	// non cluster prefixed topic validation
 	err := channels.IsValidFlowTopic(topic, c.sporkID)
 	if err != nil {
-		return NewInvalidTopicErr(topic, err)
+		return err
 	}
 	return nil
 }
+
+// validateClusterPrefixedTopic validates cluster prefixed topics.
+// Expected error returns during normal operations:
+//   - ErrActiveClusterIDsNotSet: if the cluster ID provider is not set.
+//   - ErrActiveClusterIDS: if an error is encountered while getting the active cluster IDs list. This error indicates an unexpected bug or state corruption.
+//   - channels.ErrInvalidTopic: if topic is invalid.
+//   - channels.ErrUnknownClusterID: if the topic contains a cluster ID prefix that is not in the active cluster IDs list.
+func (c *ControlMsgValidationInspector) validateClusterPrefixedTopic(from peer.ID, topic channels.Topic) error {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
```

Comment:
This lock is specifically intended to protect the `activeClusterIDs`. To better reflect its purpose, it would be clearer to rename it accordingly. Furthermore, locking the entire method just to read the active cluster IDs might be excessive. Instead, consider loading the cluster IDs once per inspected RPC (rather than per inspected topic) at an earlier point in the pipeline. For example, you could load the cluster IDs at [this location](https://github.com/onflow/flow-go/blob/66f3aaca839fffe78f9992bdb598df333b293cdb/network/p2p/inspector/validation/control_message_validation.go#L307). This approach would help reduce the performance overhead associated with acquiring and releasing locks and improve the overall efficiency of the method.